server/src/main/java/org/elasticsearch/common/util/Int3Hash.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.apache.lucene.util.hppc.BitMixer;
+import org.elasticsearch.core.Releasables;
+
+/**
+ * Specialized hash table implementation similar to BytesRefHash that maps
+ * three int values to ids. Collisions are resolved with open addressing and
+ * linear probing, growth is smooth thanks to {@link BigArrays} and capacity
+ * is always a multiple of 3 for faster identification of buckets.
+ * This class is not thread-safe.
+ */
+// IDs are internally stored as id + 1 so that 0 encodes for an empty slot
+public final class Int3Hash extends AbstractHash {
+    private IntArray keys;
+
+    // Constructor with configurable capacity and default maximum load factor.
+    public Int3Hash(long capacity, BigArrays bigArrays) {
+        this(capacity, DEFAULT_MAX_LOAD_FACTOR, bigArrays);
+    }
+
+    // Constructor with configurable capacity and load factor.
+    public Int3Hash(long capacity, float maxLoadFactor, BigArrays bigArrays) {
+        super(capacity, maxLoadFactor, bigArrays);
+        try {
+            // `super` allocates a big array so we have to `close` if we fail here or we'll leak it.
+            keys = bigArrays.newIntArray(3 * capacity, false);
+        } finally {
+            if (keys == null) {
+                close();
+            }
+        }
+    }
+
+    public int getKey1(long id) {
+        return keys.get(3 * id);
+    }
+
+    public int getKey2(long id) {
+        return keys.get(3 * id + 1);
+    }
+
+    public int getKey3(long id) {
+        return keys.get(3 * id + 2);
+    }
+
+    /**
+     * Get the id associated with <code>key</code> or -1 if the key is not contained in the hash.
+     */
+    public long find(int key1, int key2, int key3) {
+        final long slot = slot(hash(key1, key2, key3), mask);
+        for (long index = slot;; index = nextSlot(index, mask)) {
+            final long id = id(index);
+            long keyOffset = 3 * id;
+            if (id == -1 || (keys.get(keyOffset) == key1 && keys.get(keyOffset + 1) == key2 && keys.get(keyOffset + 2) == key3)) {
+                return id;
+            }
+        }
+    }
+
+    private long set(int key1, int key2, int key3, long id) {
+        assert size < maxSize;
+        final long slot = slot(hash(key1, key2, key3), mask);
+        for (long index = slot;; index = nextSlot(index, mask)) {
```

Comment:
You know I didn't write it because it has stuff like `for(;;)`. It's one of the few stylistic fingerprints we have. I'll always use a `while` loop, sometimes even `while(true)` to put you on notice "something weird is happening here". The two `;`s next to each other does the same. And it's much more your style. It's nice to have some creative freedom.

Reply:
I copied these from LongLongHash. I switched to while style in https://github.com/elastic/elasticsearch/pull/108165/commits/ca42f3b2466066b433b1144e3617eba584257666.

Reply:
I wonder who's style the funny `for` is? It's fine either way.

Reply:
I think that was Uri :) (per git blame).